docs/proposals/message-proto-design.md

```diff
@@ -0,0 +1,346 @@
+
+please see the #399 issue of kubeeedge 
+
+in addition, currently  when we use quic or websocket protocol to paas message, we need convert model message to probuf message (which defined in `github.com/kubeedge/viaduct/pkg/protos/message/message.pb.go`). and then send .  This also affects the efficiency of the transmission, rather than simply turning the model mesage into proto message。
+
+## Need to do
+
+*  Change the model message to protobuf message  directly
+*  Solve the filed index problem of `Resource`
+*  Consider compatibility  
+
+## How to do 
+
+##  Change the model message to protobuf message  directly
+
+create proto file in `model` dir to define message struct. and use `go generate`(will move to makefile) to create message.pb.go file 
+
+```
+syntax = "proto3";
+package model;
+import "google/protobuf/any.proto";
+
+message MessageHeader {
+    string ID = 1;
+    string ParentID = 2;
+    int64 Timestamp = 3;
+    bool Sync = 4;
+}
+message MessageRoute {
+    string Source = 1;
+    string Group = 2;
+    string Operation = 3;
+    google.protobuf.Any Resource = 4;
+}
+
+message Message {
+    MessageHeader Header = 1;
+    MessageRoute Router = 2;
+    google.protobuf.Any Data = 3;
+}
+```
+
+we use `Any type` to define `Resource` and `Data` field.  `Data` field is the original `Content` field.
+
+##  Solve the filed index problem of `Resource`
+
+this problem is the most important, currently, beehive define path string to stand for resource, such as `/node/{nodeID}/{namespace}/{resourceType}/{resource}`, different path stand for different resource.  but we need hardcord to parse `nodeid` ,`resourceType`,`resourceid`. It's not so extensible, so we need a more extended form. we need change `Resource` to golang `interface{}`. but in protobuf, there is not `interface` define. we need use `Any` type.
+
+```
+message MessageRoute {
+    string Source = 1;
+    string Group = 2;
+    string Operation = 3;
+    google.protobuf.Any Resource = 4;
+}
+```
+we use `github.com/gogo/protobuf`  package not offical `github.com/golang/protobuf` package . 
+
+`gogo/protobuf/types` package provides function to encodes `Any` types 
+```
+	func MarshalAny(pb proto.Message) (*Any, error) 
+	func UnmarshalAny(any *Any, pb proto.Message) error 
+	
+```
+but they can only support convert `proto.Message` into `Any` type, or convert `Any` type into `proto.Message`.  they can not convert custom type into `Any` type.
+
+so We need to support convert custom type into `Any` types, and convert `Any` types into custom types.
+
+in `pkg/core/typeurl` package of `github.com/kubeedge/beehive` project, we support custom types conversion to Any types. there are some functions:
+
+* `func RegisterType(v interface{}, args ...string)` , this function Register custom types . if we want string types can conversion to Any types. we need to add 
+ ```
+ RegisterType("", "string.content")
```

Comment:
why do we need to use .content suffix ?

Reply:
`.content` suffix is not necessary.

Reply:
do we need to register primitive types like string, bool , int etc ? or only custom types need to be registered ?

Reply:
@rohitsardesai83 , yes,  we need to register `string`, `bool` types.  because we also change  the original `Content` field type to `Any` type.

```
message Message {
    MessageHeader Header = 1;
    MessageRoute Router = 2;
    google.protobuf.Any Data = 3;
}
```
`Data` field is the original `Content` field.

in kubeedge,  there are some places where we need `string` as Content of message. such as:
https://github.com/kubeedge/kubeedge/blob/5ca1f17b87d070ccc37ea72879354355ae65857b/edge/pkg/edgehub/controller.go#L239   , so we need register it.